custom tracks import

gmngueko opened the pull request:
[this](url) extend FilogEx importer and part
of the extension import custom tracks too.
see: https://sourceforge.net/p/mytourbook/feature-requests/189/
you can test import by unzipping and importing below File
[MTB_2020-12-26.zip](https://github.com/wolfgang-ch/mytourbook/files/5743819/MTB_2020-12-26.zip)

after import the custom tracks data will be shown(summary) at this point only in the "Tour data".

Change-Id: I0b66c4d05ca13e9c0ac8f4e9bb11551f2962774b
Signed-off-by: gmngueko <gmngueko@yahoo.com>

wolfgang-ch commented:
@gmngueko 
You have not yet fully setup the Eclipse IDE because the field sorting caused a lot of changes, the code style is defined in 

http://mytourbook.sourceforge.net/mytourbook/index.php/development/dev-ide/code-style

The Eclipse code formatter do sometimes move blank lines between the class fields why you can see sometimes // empty comment lines between class fields to force formatting. I could not figure out how to prevent this, so it is recommended to review the code before it is committed.

![mt-code-formatting](https://user-images.githubusercontent.com/1283445/103205808-b758b500-48fa-11eb-88df-0932fbe4a2ad.png)

wolfgang-ch commented:
The position to display the custom data is not very good, it is somewhere between other fields, a better positon could be at the end of the fields. 

With `String.format(...)` the custom field names/values could be displayed in columns like the other field name/values

![mt-tour-data](https://user-images.githubusercontent.com/1283445/103208188-4d430e80-4900-11eb-9917-7417dd021f0f.png)

gmngueko commented:
Hello Wolfgang,

ok for the Eclipse setup. but I'am not sure we are on the same page,
so let me clarify/precise what I introduced(new from the existing MTB, "CUSTOM DATA FIELD"
is not new it is already in the current MTB !!
what I introduced is "CUSTOM DATA TRACKS" see the screen shot below, you most likely missed it
![cust-track-1](https://user-images.githubusercontent.com/50674761/103216282-f9dbbb00-4915-11eb-8db5-e6e52e997f57.jpg)
![cust-track-2](https://user-images.githubusercontent.com/50674761/103216283-fa745180-4915-11eb-9888-863adfb6f9fa.jpg)
![cust-track-3](https://user-images.githubusercontent.com/50674761/103216284-fa745180-4915-11eb-9313-22bb42b900cd.jpg)

gmngueko commented:
Sorry for the disturbance, I'am new in the project :-) .
It is most likely my fault, because I'am not sure about  what you meant by  "merge Common_PullRequests_Branch" with "main".
What I did is : 
-) pull the "Common_PullRequests_Branch" locally
-) pull the "main" branch locally too
-) take my modification I did on top of the file in my local "main" branch !!!
-) put that file straight into my locak "Common_PullRequests_Branch"
-) push it to git on the "Common_PullRequests_Branch"
-) create the Pull Request for you on that "Common_PullRequests_Branch"

So something is probably wrong in my process above :-) , can you tell me what is the correct procedure to commit actually

Thanks

wolfgang-ch commented:
> 
> 
> The position to display the custom data is not very good, it is somewhere between other fields, a better positon could be at the end of the fields.
> 
> With `String.format(...)` the custom field names/values could be displayed in columns like the other field name/values
> 
> ![mt-tour-data](https://user-images.githubusercontent.com/1283445/103208188-4d430e80-4900-11eb-9917-7417dd021f0f.png)

This was my fault, the screenshot shows the content of the tour description field and not the saved custom data

gmngueko commented:
ok, so for you the feature is good, correct ? 
I can continue in that direction ?

My next step(already implemented locally) is to incorporate the view of my "custom tracks" into the "timeslice view".
See screenshot below.
With the current approach "custom tracks" are defined for each tour (and are not a global entity like a TAG),
it is still possible to relate "custom tracks" based on the name (and not the "id" since the "id" is only locally unique for a given Tour).
So in my implementation I need to redraw the "timeslice view" when the user switch to view another tour.
![cust-track-view-1](https://user-images.githubusercontent.com/50674761/103225826-16ceb900-492b-11eb-9776-5cc2b9e58337.jpg)
![cust-track-view-2](https://user-images.githubusercontent.com/50674761/103225854-21894e00-492b-11eb-8f4d-92eb2d59cf34.jpg)

your original approach with series view and profile was that profile are pre-defined, while with my custom tracks it is no longer the case. the profile change with each tour.
It can stay that way otherwise we would need to define "custom tracks" globally with a new object that gives a definition for track/series (with Id, Name, Unit).
What is your point of view ?

thanks

wolfgang-ch commented:
> Sorry for the disturbance, I'am new in the project :-) .
> It is most likely my fault, because I'am not sure about what you meant by "merge Common_PullRequests_Branch" with "main".
> What I did is :
> -) pull the "Common_PullRequests_Branch" locally
> -) pull the "main" branch locally too
> -) take my modification I did on top of the file in my local "main" branch !!!
> -) put that file straight into my local "Common_PullRequests_Branch"
> -) push it to git on the "Common_PullRequests_Branch"
> -) create the Pull Request for you on that "Common_PullRequests_Branch"
> 
> So something is probably wrong in my process above :-) , can you tell me what is the correct procedure to commit actually

First update `Common_PullRequests_Branch` with `main`
* pull `Common_PullRequests_Branch` 
* pull `main` 
* checkout `Common_PullRequests_Branch`
* merge `main` into `Common_PullRequests_Branch`

I never do the next steps, maybe @FJBDev can describe the other steps before I try to describe them and are partly wrong :-)

FJBDev commented on bundles/net.tourbook.device/src/net/tourbook/device/sporttracks/FitLogSAXHandler.java:
```diff
@@ -448,13 +544,84 @@ private void finalizeTour() {
+               StatusUtil.log("IF (Intensity Factor)" + ": FitlogEx parse error", exc); //$NON-NLS-1$ //$NON-NLS-2$
+            }
+         }
+         if (_currentActivity.customDataFields.containsKey("Left Torque Effectiveness Avg. [%]")) { //$NON-NLS-1$
```
Per this [rule](https://rules.sonarsource.com/java/RSPEC-1192), I would recommend to define a constant instead of duplicating this literal.

FJBDev replied:
The same applies for a lot of the fields below

FJBDev commented on bundles/net.tourbook.device/src/net/tourbook/device/sporttracks/FitLogSAXHandler.java:
```diff
@@ -448,13 +544,84 @@ private void finalizeTour() {
       tourData.setDeviceTimeInterval((short) -1);
 
-      if (_currentActivity.timeSlices.isEmpty()) {
+      if (_currentActivity.timeSlices.size() == 0) {
```
Why did you revert to `.size()==0` ?
See my [PR](https://github.com/wolfgang-ch/mytourbook/pull/257) several weeks ago that explains why using isEmpty() is more performance than requesting the size per this [rule](https://rules.sonarsource.com/java/tag/clumsy/RSPEC-1155)

gmngueko replied:
@FJBDev 
I followed your recommandations and made the correction see: https://github.com/wolfgang-ch/mytourbook/pull/276

FJBDev commented:
@gmngueko 

> My next step(already implemented locally) is to incorporate the view of my "custom tracks" into the "timeslice view".

IMHO, I would recommend that you do a PR with ALL your intended modifications instead of several small PRs. For example, this PR only has the FitLogSAXHandler modifications so it's hard for me to understand the "big picture"


Good work otherwise, keep at it! :-)

>What I did is :
-) pull the "Common_PullRequests_Branch" locally
-) pull the "main" branch locally too
-) take my modification I did on top of the file in my local "main" branch !!!
-) put that file straight into my local "Common_PullRequests_Branch"
-) push it to git on the "Common_PullRequests_Branch"
-) create the Pull Request for you on that "Common_PullRequests_Branch"

You don't necessarily need to create a local "Common_PullRequests_Branch" branch.
What I do is use my local main branch that I use to create the PR against the needed branch.
For what it's worth, here is what I do:
- Do my changes in my local main
- When I'm done, update my local main with Wolfgang'a main.
- Create a PR from my main to Wolfgang's FJBDev2 branch (for you, you need to use Wolfgang's "Common_PullRequests_Branch" branch)

But it's not perfect and sometimes it doesn't work well. Several days ago, it was not working and Wolfgang and I struggled a bit because my hunch is that one of the issue is that those "side branched" (FJBDev2, Common_PullRequests_Branch...) are usually outdated and that can create conflicts with the updated main.

gmngueko commented:
@FJBDev 
thanks for the info and helps :-)
to have the big picture look at https://sourceforge.net/p/mytourbook/feature-requests/189/
(in summary I want to be able to add/define non standard data series for a Tour, "standard" series are: power, cadence, hr, etc....
non-standard series could be: estimated body temperature, or any series of data. 
If you know a bit the Garmin eco-system, you might know that you can add custom fields (connect IQ) to your recorded workout, those fields endup in the fit file and are obviously non-standard)
Here I use the ST3 FitlogEx to introduce the concept because I'am coming from ST3 and in ST3 I have a plugin which extract those non-standard workouts data from fit and save them. and I have a plugin in ST3 to export them into FitlogEx workout file which are easier to manage than fit file)
after the reading, if you look at the 2 sets of screenshots, I posted above in this thread, it will be pretty clear what I'am doing.
If after that you don't get the big picture then please let me know. I'll try to find another way to explain it.

I actually prefer smaller Pull request so that if the direction is wrong I can correct the direction early, in summary trying to be Agile.
I started with this 2 relatively small PR's because it is the basis and they are functional. And I'am new here, so not sure if the feature will actually be well received or not.( I wanted to be sure that my approach for the feature is ok)
It seems my approach is ok since it got merged but I can be wrong actually !!

FJBDev commented:
@gmngueko 

I am not familiar with the Garmin eco-system much as I have only owned Suunto devices.
After reading your feature request, I am starting to "get" the feature.
yes, your method might be a way to be Agile and you can stick with it if you think it's best, but looking at it from here, It's tough to see one file modified, then another one in another PR etc....
I guess @wolfgang-ch will be better to judge as he is the one that does the PR approval/tests.

It might also come down to personal preference. For example, I just developed the integration of Strava in MT and I did that in 1 [PR](https://github.com/wolfgang-ch/mytourbook/pull/273) to summarize and show even to the users what this complete feature is all about
Now, I had code suggestions per Wolfgang, so in the end I had to do another [PR](https://github.com/wolfgang-ch/mytourbook/pull/275) and it's OK. My point is that when I create a PR, I usually try to bring complete features all at once. And then if needed, I will do subsequent PRs to "refine" them

>And I'am new here, so not sure if the feature will actually be well received or not.
That's the same question I ask myself every time I develop a new feature :-)

Also, please look at the comments in the code.

FJBDev commented:
@gmngueko 
Will your current work on "Custom Tracks" address this [feature request](https://sourceforge.net/p/mytourbook/feature-requests/191/) ?

gmngueko commented:
yes, exactly,
fit  file importer should extract that new track "physical condition" ("condition physique" in french)
and add it as new custom track by leveraging what I did.

telemaxx commented:
hi together and this too ? https://sourceforge.net/p/mytourbook/feature-requests/157/

gmngueko commented:
Yep,
like I explain in my feature request, "custom tracks" will permit storing any non-standard track like your tss or HrTss.
I just need to refine the display of the custom tracks at this stage, the storing part is already there.
Of course it will require the ".fit" filr importer of MTB to leverage the usage of "custom tracks" for any new garmin connect IQ fields. (I had a plugin in ST3 that was doinf that, such that if a new fields appears in the fit file, it was imported into ST3).
So my work will prepare MTB to support that and then fit file decoding/import would need to be update to support this custom tracks feature.

FJBDev commented:
[Here](https://github.com/wolfgang-ch/mytourbook/blob/main/bundles/net.tourbook.device.garmin.fit/src/net/tourbook/device/garmin/fit/listeners/MesgListener_Session.java) is the FIT file import

FJBDev commented:
> 
> 
> Yep,
> like I explain in my feature request, "custom tracks" will permit storing any non-standard track like your tss or HrTss.
> I just need to refine the display of the custom tracks at this stage, the storing part is already there.
> Of course it will require the ".fit" filr importer of MTB to leverage the usage of "custom tracks" for any new garmin connect IQ fields. (I had a plugin in ST3 that was doinf that, such that if a new fields appears in the fit file, it was imported into ST3).
> So my work will prepare MTB to support that and then fit file decoding/import would need to be update to support this custom tracks feature.

However, reading the [feature request](https://sourceforge.net/p/mytourbook/feature-requests/191/), it seems that Bernard is asking for this custom data to be displayed in a graph (i.e.: Tour Chart). Is that something your feature will support as well ?

gmngueko commented:
Well,
yes down the road I plan to be able to show custom tracks on all view like standard tracks.
My plan of actions is:
1) introduce the data support in db storage (done) with view in the data view (done too)
2) support at least one way to import it (done via FitlogEx), other can follow using FitlogEx as example.
3) support timeslice view (table view in tourEditor view) (currently in the making, refining it)
4) support graph view like for any other standard track/series view (not started yet, I don't know the coed well enough at this point :-) , 3) is already chalenging :-) )

5) add metrics computing capabilities on data series (future)
6) don't differentiate custom track and standard track, make all track be defined via track definition (very futuristic :-) ), in summary make the treatment of tracks generic.

FJBDev commented:
That sounds like a great plan, awesome @gmngueko  👍

telemaxx commented:
@gmngueko , sounds cool. there are many interesting IQ data fields.
